### PR TITLE
#2616 add lookup api network helpers

### LINKED ISSUE #2616: Add patient api network helpers
## Is your feature request related to a problem? Please describe.

Add helper methods for patient API HTTP requests/responses.

## Describe the solution you'd like

Just the basic logic required to provide a minimal working demo of retrieving and displaying patient/prescriber records.

## Implementation

At this stage, simplest solution is to directly call `async` methods from the `SearchForm` component.  

## Describe alternatives you've considered

For now focusing on getting a basic working demo. Once that is given the OK, can follow up with a minor refactor and move the dispatch logic into thunks. 

## Additional context

N/A.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #2616.

Branched from #2612.

## Change summary

Just a bunch of helpers for patient API HTTP requests/responses.

Main methods are `queryPatientApi`, `queryPrescriberApi` which:

1. Transform object parameters (e.g. `completedForm`) into requests to the API endpoint.
2. Queries the server API endpoint.
3. Processes the response and returns an array of objects in a format suitable for rendering as a list.

## Testing

N/A.

### Related areas to think about

These methods are currently called directly from the `SearchForm` component. This is mostly for convenience, idea is to update this soonish such that API calls are only performed within dedicated actions/thunks.

### CODE REVIEW

File: src/utilities/network/lookupApi.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    { dob: dateOfBirth },
+    { policy_number: policyNumber },
+  ];
+  return queryParams.reduce((queryString, param) => {
+    const [[key, value]] = Object.entries(param);
+    if (!value) return queryString;
+    const paramString = `${key}=${value}`;
+    const paramSeparator = queryString.length > 0 ? SEPARATORS.PARAMETERS : SEPARATORS.QUERY_STRING;
+    return queryString + paramSeparator + paramString;
+  }, '');
```
Comment (joshxg) [inline src/utilities/network/lookupApi.js]: nice !

File: src/utilities/network/lookupApi.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  const endpoint = RESOURCES.PRESCRIBER;
+  const queryString = getQueryString(params);
+  return baseUrl + endpoint + queryString;
+};
+
+const processPatientResponse = async response => {
+  const responseJson = await response.json();
+  const { error } = responseJson;
+  if (error) throw new Error(error);
+  const patientData = responseJson.map(patient => ({
```
Comment (joshxg) [inline src/utilities/network/lookupApi.js]: Not sure how the api is working on desktop - is this the full record? so you just need to translate the remaining fields to have a saveable-record?

File: src/utilities/network/lookupApi.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    registrationCode: prescriber.code,
+  }));
+  return prescriberData;
+};
+
+export const queryPatientApi = async params => {
+  const requestUrl = getPatientRequestUrl(params);
+  try {
+    const response = await fetch(requestUrl);
+    const patientData = await processPatientResponse(response);
```
Comment (joshxg) [inline src/utilities/network/lookupApi.js]: was expecting these to be in thunks 😱

File: src/utilities/network/lookupApi.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    return patientData;
+  } catch (error) {
+    // TODO: add bugsnag.
```
Comment (joshxg) [inline src/utilities/network/lookupApi.js]: nit: I hate todo's in code :)

File: src/utilities/network/lookupApi.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    return [];
+  }
+};
+
+export const queryPrescriberApi = async params => {
+  const requestUrl = getPrescriberRequestUrl(params);
+  try {
+    const response = await fetch(requestUrl);
```
Comment (joshxg) [inline src/utilities/network/lookupApi.js]: I thought `fetch` didn't throw on error responses - so you'd need to check if the response status was in `[200, 300)`?

Also this must need authorization header?

File: src/utilities/network/lookupApi.js (hunk 3 above)
Comment (wlthomson) [reply src/utilities/network/lookupApi.js]: Yeah, I hid my shame in the related areas :laughing::

> These methods are currently called directly from the SearchForm component. This is mostly for convenience, idea is to update this soonish such that API calls are only performed within dedicated actions/thunks.

File: src/utilities/network/lookupApi.js (hunk 5 above)
Comment (wlthomson) [reply src/utilities/network/lookupApi.js]: Yeah, still very basic. Definitely needs more response handling logic. Same goes for authorisation header (but I think this requires some more work on desktop). I think both cases of for new issues (going to do an issue blitz asap once I've done a bit more testing).

File: src/utilities/network/lookupApi.js (hunk 2 above)
Comment (wlthomson) [reply src/utilities/network/lookupApi.js]: Yeah, currently quite basic, just returns the (mostly) full record. At the moment it's basically similar to the data sync receives, so I'm hoping a lot of creation logic can be reused.

File: src/utilities/network/lookupApi.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+const getQueryString = params => {
+  const {
+    firstName = '',
+    lastName = '',
+    registrationCode = '',
+    dateOfBirth = '',
+    policyNumberPerson = '',
+    policyNumberFamily = '',
+  } = params ?? {};
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/utilities/network/lookupApi.js]: Nice! 6057638c.

File: src/utilities/network/lookupApi.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
+/**
+ * mSupply Mobile
+ * Sustainable Solutions (NZ) Ltd. 2020
+ */
+
+import { SETTINGS_KEYS } from '../../settings';
+import { UIDatabase } from '../../database/index';
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/utilities/network/lookupApi.js]: 48b7ff1818ae40455fa6c480f524c1bbe3d3ef3d.

File: src/utilities/network/lookupApi.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  firstName = '',
+  lastName = '',
+  registrationCode = '',
+  dateOfBirth = '',
+  policyNumberPerson = '',
+  policyNumberFamily = '',
+} = {}) => {
+  const policyNumberSeparator =
+    policyNumberPerson && policyNumberFamily ? SEPARATORS.POLICY_NUMBER : '';
+  const policyNumber = policyNumberPerson + policyNumberSeparator + policyNumberFamily;
```
Comment (mark-prins) [inline src/utilities/network/lookupApi.js]: wrt the mSupply issue [#5680](https://github.com/sussol/msupply/pull/5860), there is a possibility that the `policyNumberPerson` would be blank in which case there shouldn't be a separator. ( another instance where the C# `Trim([char])` is great, but not available in js )

File: src/utilities/network/lookupApi.js (hunk 4 above)
Comment (mark-prins) [reply src/utilities/network/lookupApi.js]: ` // TODO: remove todos`

File: src/utilities/network/lookupApi.js (hunk 5 above)
Comment (wlthomson) [reply src/utilities/network/lookupApi.js]: Desktop PR made for authentication logic (https://github.com/sussol/msupply/pull/5905) and epic #2446 updated with issue.

File: src/utilities/network/lookupApi.js (hunk 8 above)
Comment (wlthomson) [reply src/utilities/network/lookupApi.js]: Good point. I think we've almost got agreement on the insurance policy number spec, but will wait until it's implemented just to be safe. Have updated epic with issue.

**Comment by wlthomson** [comment]
@mark-prins @joshxg All feedback noted! Have made new issues and will action in new PRs coming soon.

**Pull request merged**